sklearn/inspection/_plot/partial_dependence.py

```diff
@@ -234,6 +236,24 @@ def plot_partial_dependence(
 
         .. versionadded:: 0.24
 
+    is_categorical : list of {boolean, pair of boolean}, default=None
+        Whether each target feature in `features` is categorical or not.
+        The list should be same size as `features`. If `None`, all features
+        are assumed to be continuous.
+
+        .. versionadded:: 1.0
+
+    categorical_features : array-like of shape (n_features,) or shape \
```

Comment:
I would need some opinions of some core-dev (ping @NicolasHug @thomasjpfan). @ogrisel proposed to use the same API as for the `HistGradientBoosting`. It is indeed quite convenient and nice for the `plot_partial_dependence` function (and later on for the display).

Currently, we convert this `categorical_features` to an `is_categorical` variable that has the same shape as `features`. Indeed, `is_categorical` is passed to the `partial_dependence` function. It is kinda handy because we can parallelize easily the computation then. However, the signature of the `partial_dependence` and `plot_partial_dependence` is different. However, passing `categorical_features` to `partial_dependence` would be more complexed and would require some boilerplate validation.

I was wondering what any of you think about this issue. In short, should we go with:

1. `categorical_features` in `plot_partial_dependence` and `is_categorical` in `partial_dependence` (it might not be a huge issue since one will 95% of the time use the plotting API)
2. `categorical_features` in both `plot_partial_dependence` and `partial_dependence` (at the cost of more complexed code).

Reply:
The fact that a feature is categorical is only relevant if that feature is a target feature, right?

So I'm not sure allowing `categorical_features = some_list_of_indices` really makes sense. It seems that for both `[plot_]partial_dependence()`, we should only allow a `is_categorical` parameter that should exactly the same size as the `features` parameter. Looks like you agreed with me last year :p https://github.com/scikit-learn/scikit-learn/issues/14969#issuecomment-680726323

Reply:
> Looks like you agreed with me last year

True. Since I wrote an example :)
https://145056-843222-gh.circle-artifacts.com/0/doc/auto_examples/inspection/plot_partial_dependence.html

Since our model will have to deal with categorical data then one will need a `ColumnTransformer` and to define the name of the columns (which will even be easier the day that we can detect `category` column type). In this case, I am now under the intuition that it is more intuitive to reuse the categorical column names already defined instead of creating, each time, a list of tuples with some boolean. Indeed, it is what we are currently doing internally in `plot_partial_dependence`.

Reply:
The problem I see with `categorical_features` is that if there are categorical features but none of them are a target feature, there's no need to pass anything, but the existence of this parameter will suggest that you should. But maybe it's not too bad provided we properly document the fact that a feature is categorical only really matters for a target feature. (also, we don't support 2-ways plots for categorical features (I think?))

> It is kinda handy because we can parallelize easily the computation then

Could you detail this a bit more? Here's a silly idea: if we allow the boolean mask of `categorical_features` to have length  **either** `n_features` **or** `len(features)`, does that solve your problem? Sadly I think there's an ambiguity when `n_features == len(features)`

Reply:
> Since I wrote an example :)

Look good :)

But TBF, since as you said the `categorical_features` list must exist at some point, I would still say that my original proposal is fairly easy and natural to use. It would just be

```py
plot_partial_dependence(
...,
features=features,
is_categorical=(f in categorical_features for f in features)
...
)
```

instead of the current

```py
plot_partial_dependence(
...,
features=features,
categorical_features=categorical_features
...
)
```

This looks pretty simple and if it solves the parallelism issue, I'd go for that

Reply:
```python
is_categorical=(f in categorical_features for f in features)
```
With a mix of 1-way and 2-way, it becomes a nested `for` loop :)

> 2-ways plots for categorical features

We actually do :). This gives rise to a heatmap

Reply:
Provided that we only allow 2ways pdps when *both* features are categorical, this could just be

```py
is_categorical=((f[0] if isinstance(f, tuple) else f) in categorical_features for f in features)
```

But yeah, this becomes a bit more convoluted.